### ISSUE TITLE
BUG: stats: underflow in geom distribution

### ISSUE THREAD

**Issue by eluerken-wm** [scipy/scipy#18372]
### Describe your issue.

`stats.geom.rvs` returns negative values when `p` is small.  The support is strictly positive and should never return these values. 

When validating this in R, I am able to get valid values for substantially smaller values of `p` and this should mirror that kind of performance.

### Reproducing Code Example

```python
from scipy.stats import geom

# This returns a valid output
geom.rvs(0.0000000001, random_state = 1234) #(2125986401)
# Adding an extra zero will cause it to fail as will any smaller number
geom.rvs(0.0000000001, random_state = 1234) #(-2147483648)
```

### Error message

```shell
-2147483648
```

### SciPy/NumPy/Python version and system information

```shell
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()
1.10.1 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
  lapack:
    detection method: pkgconfig
    found: true
    include directory: c:/opt/openblas/if_32/64/include
    lib directory: c:/opt/openblas/if_32/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= PRESCOTT MAX_THREADS=4
    pc file directory: c:/opt/openblas/if_32/64/lib/pkgconfig
    version: 0.3.18
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.33
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-v_zt9njv\overlay\Lib\site-packages\pythran
    version: 0.12.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-dskf9r9k\cp310-win_amd64\build\venv\Scripts\python.exe
  version: '3.10'
```

**Comment by tupui** [comment]
Hi @eluerken-wm, thank you for reporting.

This is actually an error coming from NumPy. Behind the scene, `stats.geom.rvs` is making a call to the `geometric` distribution from NumPy.

```python
np.random.geometric(p=1e-20)
# -9223372036854775808
```

Now, this issue is only present with the legacy random number generator interface. And if you use the new `np.random.Generator`, the problem is fixed:

```python
np.random.default_rng().geometric(p=1e-20)
# 9223372036854775807
```

And you can verify this in SciPy by passing a proper `Generator` object, which is the recommended way:

```python
geom.rvs(1e-20, random_state=np.random.default_rng())
# 9223372036854775807
```

Hence, I am going to close this issue.

**Comment by tupui** [comment]
I cannot reproduce what you are getting:

```python
import scipy
np.__version__, scipy.__version__
# ('1.24.3', '1.11.0.dev0+1967.039e931')
from scipy.stats import geom
rng = np.random.default_rng()
rng.geometric(1e-20, size=15)

array([9223372036854775807, 9223372036854775807, 9223372036854775807,
       9223372036854775807, 9223372036854775807, 9223372036854775807,
       9223372036854775807, 9223372036854775807, 9223372036854775807,
       9223372036854775807, 9223372036854775807, 1786043434042018048,
       9223372036854775807, 9223372036854775807, 9223372036854775807])
```

Maybe there is more to it then: I am on a M1, could be windows things. It seemed to be linked to NumPy and the use of something else than `Generator`. That was my assumption for making the call.

Feel free to reopen and propose a workaround in SciPy if you feel otherwise.

**Comment by rkern** [comment]
It's certainly exacerbated by Windows and `RandomState`, but not exclusive to it. On 64-bit Linux (Intel), we also get negatives.

```
>>> import numpy as np

>>> rng = np.random.default_rng(0x30627150c0e97f971ae8da6864ceb85c)

>>> rng.geometric(1e-20, size=15)
array([-9223372036854775808, -9223372036854775808, -9223372036854775808,
       -9223372036854775808, -9223372036854775808,  1027431937578648064,
       -9223372036854775808, -9223372036854775808, -9223372036854775808,
       -9223372036854775808, -9223372036854775808, -9223372036854775808,
       -9223372036854775808, -9223372036854775808, -9223372036854775808])

>>> np.__version__
'1.25.0.dev0+1126.g4268399a9'
```

It might be some different behavior in what `(int64_t)ceil(...)` does on the different compilers. While there is probably a reasonable upstream fix (forcing the result in such cases to be `np.iinfo(np.int64).max`), it can still be worth doing the workaround in `geom.rvs()` as well in both `RandomState` and `Generator` cases.

**Comment by WarrenWeckesser** [comment]
> It might be some different behavior in what (int64_t)ceil(...) does on the different compilers.

Yes, that is the problem; see my recent [comment in the numpy issue](https://github.com/numpy/numpy/issues/17007#issuecomment-1524118844).

**Comment by mdhaber** [comment]
If I understand correctly, the proposed patch would change the results in this case from negative numbers, which are obviously not geometrically distributed, to positive numbers that are also not geometrically distributed, but less obviously so. If that's correct, I'd prefer not to implement this.

**Comment by eluerken-wm** [comment]
Thank you so much @tupui, @WarrenWeckesser  & @rkern!  I am sorry that I didn't reply sooner, I didn't get these notifications.  This was super informative.

I'm trying to go through the threads here, but it still seems like this issue is not fully resolved, is that fair?  I am still getting the same issues and it sounds like the proposed solution is to just flip the negative sign, is that correct?

**Comment by rkern** [comment]
We never actually give out geometrically distributed numbers because of the limitations of finite floating point and integer arithmetic. We always have a cutoff. As a matter of design of practical numerical distribution samplers, I'm very happy to have all of the probability mass in the excluded tail be assigned to the largest representable integer. This is the case even when the parameters are of typical size and only negligible amounts of probability mass are in that excluded tail; this case with extreme parameters isn't really exceptional. The overflow to negatives is a pure accident of (my) careless implementation and not a helpful feature in any way. We've already made the change for `Generator`; I think we should fix up the results that come from `RandomState` and older versions of `Generator` here.

> I am still getting the same issues and it sounds like the proposed solution is to just flip the negative sign, is that correct?

Not quite: make the result `np.iinfo(result.dtype).max`. That's typically one less than the negative of the negative integer that you get.

**Comment by mdhaber** [comment]
@rkern it looks like you reopened this. Are you willing to submit a fix or review one? If so, I'm willing to help find someone to do the other half so that we can resolve this.

**Comment by rkern** [comment]
I'll certainly review.

**Comment by mdhaber** [comment]
@OmarManzoor @dschmitz89 I thought I'd bring this issue to your attention in case you'd like to submit a PR.

**Comment by dschmitz89** [comment]
From what I understood @rkern 's proposed solution is to swap out all negative values for the maximum possible integer.

Would it not be better to raise an error or at least a `RuntimeWarning`? Otherwise, the method will silently fail without users noticing it.

**Comment by rkern** [comment]
The behavior in this edge cases used to be undefined behavior in C. One some compilers and platforms, you got the maximum integer and others you got the most negative integer. We've fixed it upstream in `Generator` to uniformly do the former. The remainder to fix here is for the codepaths still using the legacy `RandomState` which will still be platform-dependent.

I'm definitely against raising an error dependent on the values that come out the end; that's just a recipe for heisenbugs that are difficult to track down. I'm mildly okay with warnings based on the _input_ parameters where the expectation of drawing from the excluded region hits some threshold. I'm not in any particular mood to define that threshold, though.

A reminder: all of the results from our finite implementations are "failures" to draw from the true distribution. Some are just more obvious than most.

**Comment by mdhaber** [comment]
> I'm mildly okay with warnings based on the input parameters where the expectation of drawing from the excluded region hits some threshold. I'm not in any particular mood to define that threshold, though.

TLDR: For `p > 1e-17`, it's very unlikely to ever get observations clipped to the maximum `int64`; for `p < 1e-20`, it's all but guaranteed. So I'd suggest warning if `p < 1e-17`, assuming that it is relatively rare for people to be sampling with `1e-20 < p < 1e-17` and get the warning unnecessarily.

```python3
import numpy as np
max_int = np.iinfo(np.int64).max
rng = np.random.default_rng()
((rng.geometric(1e-17, size=10000000) >= max_int).any(),  # False
 (rng.geometric(1e-20, size=1) >= max_int).any())  # True
```

<details>
Given the number of observations requested, the maximum integer, and a threshold probability of getting an observation from the excluded region, the algebra gives a simple expression for the corresponding threshold for the shape `p`.

```python3
x = 1000  # maximum int 
n = 100  # number of observations from geometric distribution
pn = 0.75  # probability of at least one observation > x
p1 = 1 - (1 - pn)**(1/n)
p = 1 - p1**(1/(x + 1))
# approximation when p is very small: abs(1/(x + 1) * np.log(p1))

import numpy as np
rng = np.random.default_rng(398234982523)
any_gt_x = []
for i in range(10000):
    any_gt_x.append((rng.geometric(p, size=n) > x).any())
np.mean(any_gt_x)  # 0.7519
```

For example, if `x` is the maximum `int64`, 1 million observations are drawn, and our threshold probability is 0.1%, the threshold `p` is `2.4851459794772703e-18`. 

So if `p < 2.485e-18`, there is at least a 0.1% chance of one or more of the one million observations being clipped at the maximum integer.

But if we set `x` the maximum `int64`, the threshold for `p` is surprisingly insensitive to the number of observations (up to say, 1e15) and even the probability threshold we choosem, leading to the TLDR suggestion.

</details>

**Comment by fancidev** [comment]
I think it’s better to handle the situation based on the output rather than input. After all, for any input there is non-zero probability that the generated random number overflows; the function still has to handle this anyway. And warning on small input seems too eager to me: it’s like warning on allocating a huge block of memory that has 99% chance of failure and 1% chance of success; I think the typical behavior is to just do it and report failure if so.

While floating arithmetic is approximate and is usually addressed by rounding, the situation of integer overflow here is a bit different. *Saturation is not rounding, and truncation is not approximation*. Therefore I think it’s worth doing something in addition to saturation (which I think is a reasonable choice for the integer domain).

If a RuntimeWarning is deemed too noisy, I think it is worth at least to document the saturation behavior, which seems to me not the most obvious choice. (To me the more “Pythonic” behavior might be returning a “BigInteger” or raising a RuntimeError.)

**Comment by rkern** [comment]
I definitely agree that the saturation behavior should be documented. I still think it's acceptable to do nothing except document the behavior (Matt's findings concerning the parameter space is very informative to include, as well as an example of how one can check for saturation oneself).

But if we do want to have a warning, let me make one more pitch for doing it based on the passed input parameters rather than observing a saturated output. My main principle when it comes to warnings is that they need to be actionable for the user of the function. If the user can't change anything under their control to avoid triggering the warning, don't warn. As a kind of corollary, I think that when users make choices under their control that would lead to the warning-worthy outcomes depending on the things that are outside of their control, we should usually still warn them. Otherwise, we get heisenbugs that are hard to track down. A coder uses `geom.rvs()` with these extreme parameters and it happens to work fine with the PRNG state they tested it, but then in actual use, there are a bunch of warnings. But only sometimes.

There are two things that determine whether or not this `rvs()` method saturates the output: the parameters and the PRNG state. Practically speaking, the user only controls the parameters, not the PRNG state (certainly not in a way to control whether or not it saturates, except in contrived, party-trick scenarios). Every PRNG state is equally valid (or invalid), so I think it behooves us to think in terms of the ensemble of possible results when deciding to warn, not just when a particular PRNG state happens to push the result to the saturation limit. Our finite arithmetic always distorts the probability distribution, even aside from saturation. Under this rubric, we consider it warning-worthy when the distortion is high enough (for some value of "enough").  Even when we have very reasonable `p` values like 0.3, we're distorting the distribution. AFAICT, when `p==0.3`, our algorithm can _never_ generate the maximum `int64` integer (or even much smaller integers); it is _not_ actually true that every parameter choice has a finite chance of saturating. We just don't care about that particular distortion because the true probabilities were so low to begin with that the amount of distortion to the distribution is small. Working through the ziggurat constants, I think it's really only _possible_ to start saturating at around `p=1e-18`.

Also, more specific to this case, when `p<eps`, in addition to the saturation, we also start to observe wider and wider gaps between possible integer outputs below the saturation limit. That's an additional distortion that is only possible to detect by looking at the input parameters and not the result.

**Comment by lucascolley** [comment]
> My main principle when it comes to warnings is that they need to be actionable for the user of the function. If the user can't change anything under their control to avoid triggering the warning, don't warn. As a kind of corollary, I think that when users make choices under their control that would lead to the warning-worthy outcomes depending on the things that are outside of their control, we should usually still warn them. Otherwise, we get heisenbugs that are hard to track down. A coder uses geom.rvs() with these extreme parameters and it happens to work fine with the PRNG state they tested it, but then in actual use, there are a bunch of warnings. But only sometimes.

This is very convincing to me!

**Comment by mdhaber** [comment]
What should happen when the function can detect that there may be a problem but:
a) can't be sure, and
b) can't suggest what to do (besides solving a different problem).

For example:
```python3
import numpy as np
from scipy import linalg
rng = np.random.default_rng()
n = 1000
A = np.tril(rng.random((n, n)))
b = rng.random(n)
x = linalg.solve(A, b)
# LinAlgWarning: Ill-conditioned matrix (rcond=1.55382e-21): result may not be accurate.
```

What should this do?

**Comment by rkern** [comment]
I'm not necessarily saying that the warning message has to suggest any particular action, just that whether we issue a warning or not shouldn't hinge on something that's outside of a user's control. We often don't have nearly enough information about the user's context to suggest the right place to intervene, which might be far upstream of our library function. People solve linear systems for lots of reasons in lots of different domains, and how the ill-conditioned matrix comes into play varies. The root cause might be because the mesh in the PDE discretization is less than ideal. It might be something that could be resolved by numerical preconditioning (possibly with a different kind of solver) regardless of where the matrix comes from.

No warning message has to be comprehensive, in fact, being short and internet-searchable can be more valuable because people can find documentation and discussions that help them explore their options with higher context.

**Comment by fancidev** [comment]
@rkern Makes sense to me!

While the shape parameter has to be really small to trigger saturation (with non-negligible probability) for the geometric distribution, saturation could be more frequent for fat-tailed distributions. For example, for the Pareto distribution (whose survival probability is $x^{-\alpha}$), if $\alpha=0.001$, there is about half chance that a random number saturates to `np.inf`.

The saturation behavior for floating-point numbers could be used as a reference for integers. For most operations defined in IEEE-754, in addition to returning `+inf`, a (hardware) floating point exception is often raised unless it is suppressed (but it is often suppressed in practice).

There is no `inf` for integers, so the max integer is a natural choice for saturation.
Raising an error (or warning) on integer saturation might be a bit too intrusive, especially considering that floating point exceptions are often suppressed for this reason. Therefore I agree that documenting the saturation behavior sounds good enough (and doing more might not be desirable).

(As for the Pareto distribution, since it is known as an extreme value distribution and is often used to model excess over a threshold, the saturation behavior is particularly appropriate for such usage.)
